services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -0,0 +1,236 @@
+
+    start_periodic_time_update();
+
+    MBED_STATIC_ASSERT(sizeof(_current_time) == CURRENT_TIME_CHAR_VALUE_SIZE, "Current time characteristic value size = 10");
+
+    return bleError;
+}
+
+void CurrentTimeService::set_event_handler(EventHandler *handler) {
+    _current_time_handler = handler;
+}
+
+time_t CurrentTimeService::get_time() const
+{
+    time_t epoch_time = time(nullptr);
+
+    return epoch_time + _time_offset;
+}
+
+void CurrentTimeService::set_time(time_t host_time, uint8_t adjust_reason)
+{
+    time_t epoch_time = time(nullptr);
+
+    time_t old_time_offset = _time_offset;
+    _time_offset = host_time - epoch_time;
+
+    update_current_time_value(adjust_reason, old_time_offset - _time_offset);
+}
+
+void CurrentTimeService::update_current_time_value(const uint8_t adjust_reason, time_t time_offset_difference) {
+    time_t local_time = get_time();
+
+    struct tm *local_time_tm = localtime(&local_time);
+
+    CurrentTime current_time(local_time_tm);
+
+    current_time.adjust_reason = adjust_reason;
+
+    bool send_update = true;
+    if ((current_time.adjust_reason & (MANUAL_TIME_UPDATE | CHANGE_OF_TIME_ZONE | CHANGE_OF_DST)) == 0) {
+        if (timer.elapsed_time() < 15min) {
+            if (std::abs(time_offset_difference) <= 1) {
+                send_update = false;
+            }
+        }
+    }
+
+    if (send_update) {
+        _ble.gattServer().write(_current_time_char.getValueHandle(),
+                                reinterpret_cast<const uint8_t *>(&current_time),
+                                CURRENT_TIME_CHAR_VALUE_SIZE);
+
+        timer.reset();
+        timer.start();
+    }
+
+    start_periodic_time_update();
+}
+
+void CurrentTimeService::start_periodic_time_update() {
+    if (_event_queue_handle == 0) {
+        _event_queue_handle = _event_queue.call_in(15min, [this] {
+            _event_queue_handle = 0;
+            update_current_time_value(EXTERNAL_REFERENCE_TIME_UPDATE);
+        });
+    }
+}
+
+void CurrentTimeService::onCurrentTimeRead(GattReadAuthCallbackParams *read_request)
+{
+    time_t local_time = get_time();
+    CurrentTime local_current_time(localtime(&local_time));
+
+    if (local_current_time.valid()) {
+        _current_time = local_current_time;
+        read_request->data = reinterpret_cast<uint8_t *>(&_current_time);
+        read_request->len  = CURRENT_TIME_CHAR_VALUE_SIZE;
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    } else {
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_UNLIKELY_ERROR;
+    }
+}
+
+void CurrentTimeService::onCurrentTimeWritten(GattWriteAuthCallbackParams *write_request)
+{
+    CurrentTime input_time(write_request->data);
+
+    if (!input_time.valid()) {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_OUT_OF_RANGE;
+        return;
+    }
+
+    if (input_time.fractions256 || input_time.adjust_reason) {
```

Comment:
I don't understand why the ```|| input_time.adjust_reason``` is there, can you explain in a comment please?

Reply:
We don't know how to react for external `adjust_reason` yet. The specification doesn't help.

Reply:
I added that before `adjust_reason` was used in `set_time()`. There's no reason to ignore that field anymore. What do you think @pan- ?

Reply:
I think since the adjust reason is trasmitted in the payload it should be written as is - the spec does not mention it, so there's not reason to assume it has any different meaning

Reply:
The client must follow the spec same as servers, so they apply the same rules to the adjust reason they send